README
INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install

NAME
    PIX::Walker - Process Cisco PIX configs and 'walk' access-lists

SYNOPSIS
    PIX::Walker is an object that allows you to process PIX firewall configs and
    'walk' an access-list for matches. PIX OS versions 6 and 7 are supported.

    ** This module is still in very early development **

    'Loose' ACL matching performed. This means that you can specify as little as
    a source IP to match what line(s) that IP would match in the ACL on the
    firewall. Or you can provide every detail including source/dest IPs, ports,
    and protocol to match a specific line of an ACL. Loose matching allows you
    to see potential lines in a large ruleset that a single source or
    destination IP might match.

    More than just the first line match can be returned. If your search criteria
    can technically match multiple lines they will all be returned. This is
    useful for seeing extra matches in your ACL that might also match and can
    help you optimize your ACL.

EXAMPLE
      use PIX::Walker;

      my $config = ' ... string of full firewall config ... ';
      my $fw = new PIX::Walker($config);
      my $acl = $fw->acl('outside_access') || die("ACL does not exist");

      my $matched = 0;
      foreach my $line ($acl->lines) {
        if ($line->match(
            source => '10.0.1.100', 
            dest => '192.168.1.3', 
            port => '80', 
            protocol => 'tcp')) {
          print "Matched ACL $acl->name ($acl->elements ACE)\n" if !$matched++;
          print $line->print, "\n";
        }
      }

METHODS
    acl($name)

        Returns an PIX::Accesslist object for the ACL named by $name.

    acls()

        Returns an array of PIX::Accesslist objects for each access-list found
        in the firewall configuration. Returns undef if there is no matching
        ACL.

    alias($string)

        Returns the IP of the alias given in $alias. If no alias is found than
        the string is returned unchanged.

    findip($ip, [$trace])

        Matches the IP to an existing network-group. Does not validate it within
        any ACL. If a single group is matched a scalar is returned with the
        name, otherwise an array reference is returned containing all matches.

        * *$ip* is an IP address to look for.

        * *$trace* is an optional reference to a trace buffer. If an IP is found
        in a nested group the trace will allow you to find out where it was
        nested. See tracedump() for more information.

    findport($port, [$trace])

        Matches the PORT to an existing service-group. Does not validate it
        within any ACL. If a single group is matched a scalar is returned with
        the name, otherwise an array reference is returned containing all
        matches.

        * *$port* is the PORT to look for.

        * *$trace* is an optional reference to a trace buffer. If a PORT is
        found in a nested group the trace will allow you to find out where it
        was nested. See tracedump() for more information.

    obj($name)

        Returns an PIX::Object object for the object-group that matches the
        $name given.

    tracedump($trace)

        Prints out the trace dump given. This will allow you to see where IP's
        and PORT's are being matched within their object-groups even if they are
        nested.

                    $matched = $fw->findip($ip, $trace);
                    $fw->tracedump($trace);

AUTHOR
        Jason Morriss, "<lifo at liche.net>"

BUGS
        Please report any bugs or feature requests to "bug-pix-walker at
        rt.cpan.org", or through the web interface at
        <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=PIX-Walker>. I will be
        notified, and then you'll automatically be notified of progress on your
        bug as I make changes.

SUPPORT
        This POD document is the only support you will receive on this module.

ACKNOWLEDGEMENTS
        Peter Vargo - For pushing me to make this module and for supplying me
        with endless ideas.

COPYRIGHT & LICENSE
        Copyright 2006 Jason Morriss, all rights reserved.

        This program is free software; you can redistribute it and/or modify it
        under the same terms as Perl itself.



Changes
Revision history for PIX-Walker

1.03	June 13, 2006
	Added some methods to PIX::Accesslist::Line for fetching
	the logical elements of a line as strings (protocol, source, 
	dest, etc..)

	Normalized the PIX::Walker standard set of port aliases.

	Added PIX::Object::icmp_type -- Doesn't actually do anything
	though. It's a useless object type to me at the moment. This
	is included so configs with object-groups of icmp-type will 
	be processed correctly.

1.02    June 07, 2006
        First version, released on an unsuspecting world.



lib/PIX/Object.pm
package PIX::Object;
# Factory class for the various object-groups found in a PIX config

use strict;
use warnings;

use Carp;

our $VERSION = '1.00';

sub new {
	my $proto = shift;
	my $class = ref($proto) || $proto;
	my $self = { };
	my ($type, $name, $config, $walker) = @_;
	croak("Must provide the object-group type, name and config block") unless ($type and $name and $config);

	$class .= "::" . lc $type;
	$class =~ tr/-/_/;
	eval "require $class";
	if ($@) {
		die("Object subclass '$class' has compile time errors:\n$@\n");
	} 

	$self->{class} = $class;
	$self->{name} = $name;
	$self->{type} = $type;
	$self->{config} = [ @$config ];
	$self->{config_block} = [ @$config ];
	$self->{walker} = $walker;

	bless($self, $class);
	$self->_init;

	return $self;
}

sub type { $_[0]->{type} }
sub name { $_[0]->{name} }
sub list { }
sub first { ($_[0]->list)[0] }

sub alias {
	my $self = shift;
	my $string = shift;
	return defined $self->{walker} ? $self->{walker}->alias($string) : $string;
}

sub _init {
	my $self = shift;

	if (@{$self->{config_block}}[0] !~ /^object-group \S+ \S+/i) {
		carp("Invalid config block passed to $self->{class}");
		return undef;
	}
}

sub _nextline { shift @{$_[0]->{config_block}} }
sub _rewind { unshift @{$_[0]->{config_block}}, $_[1] }

1;


lib/PIX/Accesslist.pm
package PIX::Accesslist;

use strict;
use warnings;

use Carp;
#use FWcommon;
use PIX::Accesslist::Line;

our $VERSION = '1.00';

=pod

=head1 NAME

PIX::Accesslist - Accesslist object for use with PIX::Walker

=head1 SYNOPSIS

PIX::Accesslist is used by PIX::Walker to hold an ACL from a PIX firewall. 
This allows you to programmatically step through an ACL and match lines
to certain criteria.

See B<PIX::Walker> for an example.

  $acl = new PIX::Accesslist($name, $acl_conf, $walker);			


=head1 METHODS

=cut

sub new {
	my $proto = shift;
	my $class = ref($proto) || $proto;
	my $self = { };
	my ($name, $config, $walker) = @_;
	croak("Must provide the access-list name and config block") unless ($name and $config);

	$self->{class} = $class;
	$self->{name} = $name;
	$self->{config} = [ @$config ];
	$self->{config_block} = [ @$config ];
	$self->{walker} = $walker;
	$self->{acl} = [];


	bless($self, $class);
	$self->_init;

	return $self;
}

sub _init {
	my $self = shift;

	if (@{$self->{config_block}}[0] !~ /^access-list \S+ \S+/i) {
		carp("Invalid config block passed to $self->{class}");
		return undef;
	}

	$self->{unnamed_any} = 0;
	$self->{unnamed_host} = 0;
	$self->{unnamed_net} = 0;
	$self->{unnamed_proto} = 0;
	$self->{unnamed_service} = 0;

	my $idx = 0;
	my ($word, $next);
	while (defined(my $line = $self->_nextline)) {
		$idx++;
		$line =~ s/^access-list $self->{name}(?: extended)? //;
		next if $line =~ /^remark/;			# ignore remarks
		next unless $line =~ s/^(permit|deny)//;	# strip off action
		my $action = lc $1;
		my $proto = $self->_getproto(\$line);
		my $source = $self->_getnetwork(\$line);
		my $sport = $self->_getports(\$line, $proto);		# will be undef if there's no port or service object-group
		my $dest = $self->_getnetwork(\$line);
		my $dport = $self->_getports(\$line, $proto);		# ... 

		my $o = new PIX::Accesslist::Line($action, $proto, $source, $sport, $dest, $dport, $idx);
		push(@{$self->{acl}}, $o);
	}
}

# returns the next network from the current line. $line is a ref
sub _getnetwork {
	my ($self, $line) = @_;
	croak("\$line must be a reference") unless ref $line;
	my $net;

	my $word = $self->_nextword($line);
	if ($word eq 'object-group') {
		$net = $self->{walker}->obj( $self->_nextword($line) );

	} elsif ($word eq 'any') {
		my $name = 'unnamed_any_'.(++$self->{unnamed_any});
		my $conf = [ 
			"object-group network $name", 
			"network-object 0.0.0.0 0.0.0.0" 
		];
		$net = new PIX::Object('network', $name, $conf, $self->{walker});

	} elsif ($word eq 'host') {
		my $ip = $self->_nextword($line);
		my $name = 'unnamed_host_'.(++$self->{unnamed_host});
		my $conf = [ 
			"object-group network $name", 
			"network-object " . $self->{walker}->alias($ip) . " 255.255.255.255" 
		];
		$net = new PIX::Object('network', $name, $conf, $self->{walker});

	} elsif (($word =~ /^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}/) or ($word ne $self->{walker}->alias($word))) {
		my $name = 'unnamed_net_'.(++$self->{unnamed_net});
		my $conf = [ 
			"object-group network $name", 
			"network-object " . $self->{walker}->alias($word) . " " . $self->_nextword($line) 
		];
		$net = new PIX::Object('network', $name, $conf, $self->{walker});

	} else {
		print "** Unknown network: $word $$line\n";
	}
	return $net;
}

# returns the protocol from the current line. $line is a ref
sub _getproto {
	my ($self, $line) = @_;
	croak("\$line must be a reference") unless ref $line;
	my $proto;

	my $word = $self->_nextword($line);
	if ($word eq 'object-group') {
		$proto = $self->{walker}->obj( $self->_nextword($line) );
	} else {
		my $name = 'unnamed_proto_'.(++$self->{unnamed_proto});
		my $conf = [ "object-group protocol $name", "protocol-object $word" ];
		$proto = new PIX::Object('protocol', $name, $conf, $self->{walker});
	}

	return $proto;
}

# returns the next service port(s) from the current line. $line is a ref
sub _getports {
	my ($self, $line, $proto) = @_;
	croak("\$line must be a reference") unless ref $line;
	my $port;

	my $word = $self->_nextword($line) || return undef;
	if ($word eq 'object-group') {
		my $word2 = $self->_nextword($line);
		my $obj = $self->{walker}->obj($word2);
		$port = (defined($obj) and $obj->type eq 'service') ? $obj : undef;
		# put the previous words back in the line, since it's going to be a valid network object-group
		if (!$port) {
			$self->_rewindword($word2, $line);
			$self->_rewindword($word, $line);
		}
	} elsif ($word eq 'eq') {
		$word = $self->_nextword($line);
		my $name = 'unnamed_service_'.(++$self->{unnamed_service});
		my $conf = [ "object-group service $name", "port-object eq $word" ];
		$port = new PIX::Object('service', $name, $conf, $self->{walker});
	} elsif ($word eq 'gt') {
		$word = $self->_nextword($line);
		my $name = 'unnamed_service_'.(++$self->{unnamed_service});
		my $conf = [ "object-group service $name", "port-object gt $word" ];
		$port = new PIX::Object('service', $name, $conf, $self->{walker});
	} elsif ($word eq 'lt') {
		$word = $self->_nextword($line);
		my $name = 'unnamed_service_'.(++$self->{unnamed_service});
		my $conf = [ "object-group service $name", "port-object lt $word" ];
		$port = new PIX::Object('service', $name, $conf, $self->{walker});
	} elsif ($word eq 'range') {
		$word = $self->_nextword($line);
		my $word2 = $self->_nextword($line);
		my $name = 'unnamed_service_'.(++$self->{unnamed_service});
		my $conf = [ "object-group service $name", "port-object range $word $word2" ];
		$port = new PIX::Object('service', $name, $conf, $self->{walker});
	} elsif ($word eq 'neg') {
		$word = $self->_nextword($line);
		my $name = 'unnamed_service_'.(++$self->{unnamed_service});
		my $conf = [ "object-group service $name", "port-object neg $word" ];
		$port = new PIX::Object('service', $name, $conf, $self->{walker});
	} else {	# any other values (eg: 'log') are ignored
		$self->_rewindword($word, $line);
		return undef;
	}

	# save the newly created service group to the main walker object
	if (defined $port and !defined $self->{walker}->obj($port->name)) {
		$self->{walker}{objects}{$port->name} = $port;
	}

	return $port;
}

=item B<elements()>

=over

Returns the total elements (ACE) in the access-list.
B<Note:> It's not wise to call this over and over again. Store the result
in a variable and use that variable if you need to use this result in multiple
places.

=back

=cut
sub elements {
	my $self = shift;
	my $total = 0;
	$total += $_->elements for $self->lines;
	return $total;
}

=item B<lines()>

=over

Returns all lines of the ACL. Each line is an B<PIX::Accesslist::Line> object.

=back

=cut
sub lines { @{$_[0]->{acl}} }

=item B<name()>

=over

Returns the name of the ACL

=back

=cut
sub name { $_[0]->{name} }

=item B<print()>

=over

Pretty prints the ACL. Tries to make it easy to read. If object-group's are used
the names are printed instead of IP's if more than a single IP is present for a line.

  1)  permit (tcp)   192.168.0.0/24 -> 0.0.0.0/0 [Web_Services_tcp: 80,443]
  10) deny   (ip)    0.0.0.0/0 -> 0.0.0.0/0

=back

=cut
sub print {
	my $self = shift;
	my $output = "----- Access-list $self->{name} -----\n";
	$output .= $_->print . "\n" for $self->lines;
	return $output;
}

# $line is a ref to a scalar string. The word returned is removed from the string
sub _nextword { (${$_[1]} =~ s/^\s*(\S+)\s*//) ? $1 : undef; }
sub _nextline { shift @{$_[0]->{config_block}} }
sub _reset { $_[0]->{config_block} = $_[0]->{config} }
sub _rewind { unshift @{$_[0]->{config_block}}, $_[1] }
sub _rewindword { ${$_[2]} = $_[1] . " " . ${$_[2]} }

1;

__DATA__

=head1 AUTHOR

Jason Morriss, C<< <lifo at liche.net> >>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-pix-walker at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=PIX-Walker>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SUPPORT

    perldoc PIX::Walker
    perldoc PIX::Accesslist
    perldoc PIX::Accesslist::Line

=head1 COPYRIGHT & LICENSE

Copyright 2006 Jason Morriss, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut


lib/PIX/Object/network.pm
package PIX::Object::network;

use strict;
use warnings;
use base qw( PIX::Object );

our $VERSION = '1.00';

sub _init {
	my $self = shift;
	$self->SUPER::_init;

	$self->{debug} = 0;

	$self->{networks} = [];				# networks directly defined in this object group
	$self->{groups} = [];				# other groups that this object points to for networks

	$self->_nextline;				# remove the first line 'object-group blah'
	while (defined(my $line = $self->_nextline)) {
		my ($ip, $mask);
#		print "$self->{name}: $line\n";
		if ($line =~ /\s*network-object (\S+) (\S+)/) {
			($ip, $mask) = ($1, $2);
			if ($ip eq 'host') {
				$ip = $mask;
				$mask = '255.255.255.255';
			}
			$ip = $self->alias($ip);
			$self->add($ip, $mask);
		} elsif ($line =~ /^\s*group-object (\S+)/) {
			push(@{$self->{groups}}, $1);
		} elsif ($line =~ /^\s+description (.+)/) {
			$self->{desc} = $1;
		} else {
			warn "$self->{name}: Unknown network object line: $line\n"; 
		}
	}
#	use Data::Dumper;
#	print "$self->{name}:" . Dumper($self->{networks});
}

# searches the networks within our group for the IP given. 
# Deligates out to nested groups and maintains the state of the trace.
sub matchip {
	my ($self, $ip, $trace) = @_;
	my $ipint = ip2int($ip);
	my $found = undef;
	my $idx = 0;

	# first search our defined networks.
#	print "searching networks in $self->{name} ...\n" if $self->{debug};
	foreach my $net (@{$self->{networks}}) {
		$idx++;
		if ($ipint >= $net->[0] and $ipint <= $net->[1]) {
			push(@$trace, $self->{name}, $idx, int2ip($net->[0]) . "/" . $net->[2]) if defined $trace;
			return $self->{name};
		}
	}

	# search all nested groups, if any.
	$idx = 0;
#	print "searching groups in $self->{name} (" . (join(',', @{$self->{groups}})) . ")...\n" if $self->{debug} and scalar @{$self->{groups}};
	foreach my $name (@{$self->{groups}}) {
		$idx++;
		my $grp = $self->{walker}->obj($name) || next;
		my $localtrace = [ $self->{name}, 0, 0 ];
		next unless $grp->type eq 'network';
		my $found = $grp->matchip($ip, $localtrace);
		if ($found) {
			push(@$trace, @$localtrace) if defined $trace;
			return $grp->{name};
		}
	}	
}

sub networks { return wantarray ? @{$_[0]->{networks}} : $_[0]->{networks} }
sub groups { return wantarray ? @{$_[0]->{groups}} : $_[0]->{groups} }

sub add {
	my ($self, $ip, $mask) = @_;
	my ($first, $last, $bits);
	$bits = ipnumbits($mask);
	$first = ip2int($ip);
	$last = ip2int(ipbroadcast($first, $bits));
	push(@{$self->{networks}}, [ $first, $last, $bits ]);
}

# returns a list of networks
sub list {
	my $self = shift;
	my @list = ();
	for (my $i=0; $i<@{$self->{networks}}; $i++) {
		push(@list, int2ip($self->{networks}[$i][0]) . ($self->{networks}[$i][2] ne '32' ? '/' . $self->{networks}[$i][2] : ''));
	}
	foreach my $name ($self->groups) {
		my $grp = $self->{walker}->obj($name) || next;
		push(@list, $grp->list);
	}
	return @list;
}

sub ip2int {
	my ($ip, $port) = split(/:/, shift, 2);		# strip off any port if it's present
	my ($i1,$i2,$i3,$i4) = split(/\./, $ip);
	return ($i4) | ($i3 << 8) | ($i2 << 16) | ($i1 << 24);
}

sub int2ip {
	my $num = shift;
	return join(".", 
		($num & 0xFF000000) >> 24,
		($num & 0x00FF0000) >> 16,
		($num & 0x0000FF00) >> 8,
		($num & 0x000000FF)
	);
}

sub ipnetmask {
	my $bits = shift;
	return '0.0.0.0' unless $bits;
	my $num = 0xFFFFFFFF;
	my $mask = ($num >> (32 - $bits)) << (32 - $bits);
	return int2ip($mask);
}

sub ipwildmask {
	my $bits = shift;
	my $num = ip2int( ipnetmask($bits) );
	$num = $num ^ 0xFFFFFFFF;
	return int2ip($num);
}

sub ipbroadcast {
	my ($num, $bits) = @_;
	my @ip = split(/\./, int2ip($num));
	my @wc = split(/\./, ipwildmask($bits));
	my $bc = "";
	for (my $i=0; $i < 4; $i++) { $ip[$i] += $wc[$i]; }
	return join(".",@ip);
}

sub ipnumbits {
	my ($mask) = @_;
	my $bits = unpack('B32', pack('N', ip2int($mask)));
	return scalar grep { $_ eq '1' } split(//, $bits);
}


1;


lib/PIX/Object/icmp_type.pm
package PIX::Object::icmp_type;

use strict;
use warnings;
use base qw( PIX::Object );

our $VERSION = '1.00';

sub _init {
	my $self = shift;
	$self->SUPER::_init;

	$self->{debug} = 0;

	$self->{icmptypes} = [];			# icmp-types directly defined in this object group
	$self->{groups} = [];				# other groups that this object points to for icmptypes

	# ... currently objects of this type are not processed ...
}

sub icmptypes { return wantarray ? @{$_[0]->{icmptypes}} : $_[0]->{icmptypes} }
sub groups { return wantarray ? @{$_[0]->{groups}} : $_[0]->{groups} }

sub add {
	my ($self, $i) = @_;
	push(@{$self->{icmptypes}}, $i);
}

sub list {
	my $self = shift;
	my @list = @{$self->{icmptypes}};
	foreach my $name ($self->groups) {
		my $grp = $self->{walker}->obj($name) || next;
		push(@list, $grp->list);
	}
	return @list;
}


1;


lib/PIX/Walker.pm
package PIX::Walker;

use strict;
use warnings;

use Carp;
use PIX::Object;
use PIX::Accesslist;

BEGIN {
	use Exporter();

	our ($VERSION, @ISA, @EXPORT, @EXPORT_OK, %EXPORT_TAGS);

	$VERSION = '1.03';

	@ISA = qw(Exporter);
	@EXPORT = qw();
	@EXPORT_OK = qw();
	%EXPORT_TAGS = ();

}

=pod

=head1 NAME

PIX::Walker - Process Cisco PIX configs and 'walk' access-lists

=head1 SYNOPSIS

PIX::Walker is an object that allows you to process PIX firewall configs and 'walk' an access-list for matches.
PIX OS versions 6 and 7 are supported.

B<** This module is still in very early development **>

'Loose' ACL matching performed. This means that you can specify as little as a source IP to match what line(s) that IP
would match in the ACL on the firewall. Or you can provide every detail including source/dest IPs, ports, and protocol to 
match a specific line of an ACL. Loose matching allows you to see potential lines in a large ruleset that a single source or
destination IP might match.

More than just the first line match can be returned. 
If your search criteria can technically match multiple lines they will all be returned.
This is useful for seeing extra matches in your ACL that might also match and can help you optimize your ACL.

=head1 EXAMPLE

  use PIX::Walker;

  my $config = ' ... string of full firewall config ... ';
  my $fw = new PIX::Walker($config);
  my $acl = $fw->acl('outside_access') || die("ACL does not exist");

  my $matched = 0;
  foreach my $line ($acl->lines) {
    if ($line->match(
        source => '10.0.1.100', 
        dest => '192.168.1.3', 
        port => '80', 
        protocol => 'tcp')) {
      print "Matched ACL $acl->name ($acl->elements ACE)\n" if !$matched++;
      print $line->print, "\n";
    }
  }

=head1 METHODS

=cut

sub new {
	my $proto = shift;
	my $class = ref($proto) || $proto;
	my $self = { debug => 0 };
	my ($fw_config) = @_;
	my $conf;
	croak("Must provide firewall configuration") unless $fw_config;

	bless($self, $class);

	$self->{firewall} = '';

	my ($file, $host);

	$file = (split(/\n/, $fw_config, 2))[0];
	$host = (split(/\./, $file, 2))[0];

	if (-f $file) {
#		print "Reading from file\n";
		open(F, "<$file") or die("Error opening file for reading: $!");
		$conf = join('', <F>);
		close(F);
	} elsif (-f "$file.conf") {
#		print "Reading from conf file\n";
		open(F, "<$file.conf") or die("Error opening file for reading: $!");
		$conf = join('', <F>);
		close(F);
	} elsif (-f "$host") {
		print "Reading from host file\n";
#		open(F, "<$host") or die("Error opening file for reading: $!");
		$conf = join('', <F>);
		close(F);
	} elsif (-f "$host.conf") {
#		print "Reading from host conf file\n";
		open(F, "<$host.conf") or die("Error opening file for reading: $!");
		$conf = join('', <F>);
		close(F);
	} else {
		$conf = $fw_config;
	}
	croak("No firewall configuration found") unless $conf;
	$self->{config} = [ split(/\n/, $conf) ];
	$self->{config_block} = [ split(/\n/, $conf) ];

	$self->_init;
	$self->_process;

	return $self;
}

sub _init {
	my $self = shift;
	$self->{objects} = {};
	$self->{acls} = {};
	$self->{alias} = {};
	$self->{ports} = {
		# insert static entries here...
		'imap4'			=> '143',
		'h323'			=> '1720',
		'sqlnet'		=> '1521',
		'pcanywhere-data'	=> '5631',
		'pcanywhere-status'	=> '5632',
		'citrix-ica'		=> '1494',

		# cisco PIX defined
		'aol'			=> '5190',
		'bgp'			=> '179',
		'biff'			=> '512',
		'bootpc'		=> '68',
		'bootps'		=> '67',
		'chargen'		=> '19',
		'cmd'			=> '514',
		'daytime'		=> '13',
		'discard'		=> '9',
		'domain'		=> '53',
		'dnsix'			=> '195',
		'echo'			=> '7',
		'exec'			=> '512',
		'finger'		=> '79',
		'ftp'			=> '21',
		'ftp-data'		=> '20',
		'gopher'		=> '70',
		'hostname'		=> '101',
		'https'			=> '443',
		'nameserver'		=> '42',
		'ident'			=> '113',
		'irc'			=> '194',
		'isakmp'		=> '500',
		'klogin'		=> '543',
		'kshell'		=> '544',
		'ldap'			=> '389',
		'ldaps'			=> '636',
		'lotusnotes'		=> '1352',
		'lpd'			=> '515',
		'login'			=> '513',
		'mobile-ip'		=> '434',
		'netbios-ns'		=> '137',
		'netbios-dgm'		=> '138',
		'netbios-ssn'		=> '139',
		'nntp'			=> '119',
		'ntp'			=> '123',
		'pim-auto-rp'		=> '496',
		'pop2'			=> '109',
		'pop3'			=> '110',
		'pptp'			=> '1723',
		'rip'			=> '520',
		'rtsp'			=> '554',
		'sip'			=> '5060',
		'smtp'			=> '25',
		'snmp'			=> '161',
		'snmptrap'		=> '162',
		'ssh'			=> '22',
		'sunrpc'		=> '111',
		'syslog'		=> '514',
		'tacacs'		=> '49',
		'talk'			=> '517',
		'telnet'		=> '23',
		'tftp'			=> '69',
		'time'			=> '37',
		'uucp'			=> '540',
		'who'			=> '513',
		'whois'			=> '43',
		'www'			=> '80',
		'xdmcp'			=> '177',
	};

	# look for services files (nmap is better) and build a translation table.
	# this reads ALL the files listed and merges the results into a single hash lookup table.
	# the first name-to-port lookup found is used and is not overwritten
#	my @files = qw( ./port-numbers /usr/local/share/nmap/nmap-services /usr/share/nmap/nmap-services /etc/services );
	my @files = qw( /usr/local/share/nmap/nmap-services /usr/share/nmap/nmap-services /etc/services );
	while ( defined(my $file = shift @files)) {
		next unless -f $file;
		open(F, "<$file") or next;
		while (defined(my $line = <F>)) {
			$line =~ s/^\s+//;
			$line =~ s/\s+$//;
			next if $line eq '' or $line =~ /^#/;
			my ($name, $str) = split(/\s+/, $line);
			my $port = (split(/\//, $str))[0];
			$self->{ports}{$name} = $port unless exists $self->{ports}{$name};
		}
		close(F);
#		last;	# uncomment this to only read the FIRST file found
	}
}

# INTERNAL: processes the configuration and breaks things apart into different bits
sub _process {
	my $self = shift;

	# continue until all config lines are gone
	while (defined(my $line = $self->_nextline)) {
		if ($line =~ /^object-group (\S+) (\S+)/i) {
			my ($type, $name) = ($1,$2);
			my $conf = [ $line ];
			$line = $self->_nextline;
			while (defined $line && $line =~ /^\s*(\w+-object|desc)/) {
				push(@$conf, $line);
				$line = $self->_nextline;
			}
			$self->_rewind($line);		# rewind 1 line so we don't skip past it on the next iteration
			$self->{objects}{$name} = new PIX::Object($type, $name, $conf, $self);
		} elsif ($line =~ /^access-list (\S+)/) {
			my $name = $1;
			next if $name eq 'compiled';
			my $conf = [ $line ];
			$line = $self->_nextline;
			while (defined $line && $line =~ /^access-list $name/) {
				push(@$conf, $line);
				$line = $self->_nextline;
			}
			$self->_rewind($line);
			$self->{acls}{$name} = new PIX::Accesslist($name, $conf, $self);			

		} elsif ($line =~ /^name (\S+) (.+)/) {
			$self->{alias}{$2} = $1;
		}
	}

}

=item B<acl($name)>

=over

Returns an B<PIX::Accesslist> object for the ACL named by $name.

=back

=cut
sub acl {
	my $self = shift;
	my $name = shift;
	return exists $self->{acls}{$name} ? $self->{acls}{$name} : undef;
}

=item B<acls()>

=over

Returns an array of PIX::Accesslist objects for each access-list 
found in the firewall configuration. Returns undef if there is no
matching ACL.

=back

=cut
sub acls { keys %{$_[0]->{acls}} }

=item B<alias($string)>

=over

Returns the IP of the alias given in $alias. If no alias is found than the
string is returned unchanged.

=back

=cut
sub alias {
	my $self = shift;
	my $string = shift;
	return exists $self->{alias}{$string} ? $self->{alias}{$string} : $string;
}

=item B<findip($ip, [$trace])>

=over

Matches the IP to an existing network-group. Does not validate it within any ACL.
If a single group is matched a scalar is returned with the name, otherwise an 
array reference is returned containing all matches.

* I<$ip> is an IP address to look for.

* I<$trace> is an optional reference to a trace buffer. 
If an IP is found in a nested group the trace will allow you to find out where 
it was nested. See L<tracedump()> for more information.

=back

=cut
sub findip {
	my ($self, $ip, $trace) = @_;
	my $found = [];

	foreach my $obj (keys %{$self->{objects}}) {
		my $grp = $self->{objects}{$obj};
		next unless $grp->type eq 'network';	# we only care about network groups
		my $localtrace = [];
		my $match = $grp->matchip($ip, $localtrace);
		if ($match) {
			push(@$trace, $localtrace) if defined $trace;
			push(@$found, $match);
		}
	}
	if (scalar @$found) {
		my %u;
		my @uniq = grep { !$u{$_}++ } sort @$found;
		return (scalar @uniq == 1) ? $uniq[0] : \@uniq;
	}
	return undef;
}

=item B<findport($port, [$trace])>

=over

Matches the PORT to an existing service-group. Does not validate it within any ACL.
If a single group is matched a scalar is returned with the name, otherwise an 
array reference is returned containing all matches.

* I<$port> is the PORT to look for.

* I<$trace> is an optional reference to a trace buffer. 
If a PORT is found in a nested group the trace will allow you to find out where 
it was nested. See L<tracedump()> for more information.

=back

=cut
sub findport {
	my ($self, $port, $trace) = @_;
	my $found = [];

	foreach my $obj (keys %{$self->{objects}}) {
		my $grp = $self->{objects}{$obj};
		next unless $grp->type eq 'service';	# we only care about service groups
		my $localtrace = [];
		my $match = $grp->matchport($port, $localtrace);
		if ($match) {
			push(@$trace, $localtrace) if defined $trace;
			push(@$found, $match);
		}
	}
	if (scalar @$found) {
		my %u;
		my @uniq = grep { !$u{$_}++ } sort @$found;
		return (scalar @uniq == 1) ? $uniq[0] : \@uniq;
	}
	return undef;
}

=item B<obj($name)>

=over

Returns an B<PIX::Object> object for the object-group that matches the $name given.

=back

=cut
sub obj {
	my $self = shift;
	my $name = shift;
	return exists $self->{objects}{$name} ? $self->{objects}{$name} : undef;
}


sub portnum { 
	my ($self, $port) = @_;
	return $port if $port =~ /^\d+$/;
	# using die() below is a bit harsh but I don't have a better way to deal with it for now.
	return exists $self->{ports}{$port} ? $self->{ports}{$port} : die("Unknown port name '$port'");
}

=item B<tracedump($trace)>

=over

Prints out the trace dump given. This will allow you to see where IP's and PORT's are being 
matched within their object-groups even if they are nested.

=over

	$matched = $fw->findip($ip, $trace);
	$fw->tracedump($trace);

=back

=cut
sub tracedump {
	my ($self, $trace) = @_;
	return '' unless defined $trace;
	print "\nMatch Trace: \n" if @$trace;
#	use Data::Dumper; print Dumper($trace); return;
	foreach my $tr (@$trace) {
		my $idx = 0;
		for (my $i=0; $i<@$tr; $i=$i+3) {
			my ($name, $line, $extra) = @$tr[$i..$i+2];
#			print "\t"x($idx++) . $name;
			print " -> " if $idx++;
			print $name;
			print " (match: $extra)" if $extra;
			print " (idx: $line)" if $line;
#			print "\n";
		}
		print "\n";
	}
	print "\n";
}

sub _nextline { shift @{$_[0]->{config_block}} }
sub _rewind { unshift @{$_[0]->{config_block}}, $_[1] }
sub _reset { $_[0]->{config_block} = $_[0]->{config} }

sub total_config_lines {
	my $self = shift;
	return 0 unless defined $self->{config};
	return scalar @{$self->{config}};
}
sub total_network_objects { my $self=shift; return scalar grep { $self->{objects}{$_}->{class} =~ /network$/ } keys %{$self->{objects}} }
sub total_service_objects { my $self=shift; return scalar grep { $self->{objects}{$_}->{class} =~ /service$/ } keys %{$self->{objects}} }
sub total_protocol_objects { my $self=shift; return scalar grep { $self->{objects}{$_}->{class} =~ /protocol$/ } keys %{$self->{objects}} }
sub total_icmp_type_objects { my $self=shift; return scalar grep { $self->{objects}{$_}->{class} =~ /icmp_type$/ } keys %{$self->{objects}} }
sub total_object_groups { return scalar keys %{$_[0]->{objects}} }
##sub total_acl_lines { return scalar @{$_[0]->{acl}} }

1;

__DATA__

=head1 AUTHOR

Jason Morriss, C<< <lifo at liche.net> >>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-pix-walker at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=PIX-Walker>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SUPPORT

This POD document is the only support you will receive on this module.

=head1 ACKNOWLEDGEMENTS

B<Peter Vargo> - For pushing me to make this module and for supplying me with endless ideas.

=head1 COPYRIGHT & LICENSE

Copyright 2006 Jason Morriss, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut
